File: parser.py
#!/usr/bin/python
import re
import os
import locale

locale.setlocale( locale.LC_ALL, 'en_US.UTF-8' ) 

'''
 Define 2 dictionaries to save the data,
 dic is dictionary of artist: works, 
 dicValue is dictionary of artist: totalValue.
'''
dic = {}	
dicValue = {}	

''' 
 Parse the documents,
 open the documents in the loop
'''
Path = "data/2017-12-20/"
filelist = os.listdir(Path)
for file in filelist:	
    if file.endswith(".html") and file.startswith("lot"): 
		fh = open(Path+file, "r")
		num = 1
		line = fh.readline()
        # search line by line
		while line:

			# ''' search the artist name in a html file '''
			if num == 1:
				reobj = re.search( r'<h3 class=\'artist\'>(.*?)( \(.*\))*</h3>', line)
				if reobj:
					artist = '\''+reobj.group(1)+'\''
					# initialize the dictionaries of the artist
					if artist not in dic:
						# the artist's works are saved as a list 
						dic[artist] = []	
						dicValue[artist] = 0
					num += 1

			# ''' search the title of the work in a html file '''
			elif num == 2:
				reobj = re.search( r'<h3>(.*?)</h3>', line)
				if reobj:
					title = '\''+reobj.group(1)+'\''
					num += 1

			# '''
			#  search the price of the work in a html file
			#  each work is saved as a dictionary
			# '''
			elif num == 3:
				reobj = re.search( r'<div><span class=\'currency\'>(.*?)</span><span>(.*?)</span></div>', line)
				if reobj:
					currency = '\''+reobj.group(1)+'\''
					# convert string to integer
					value = locale.atoi(reobj.group(2))	
					# convert GBP to USD
					if currency == '\'GBP\'':	
						value *= 1.34
						currency = '\'USD\''
					# calculate the totalValue
					dicValue[artist] += value;	
					amount = '\''+"{:,}".format(value)+'\''
					dic[artist].append({'title': title, 'currency': currency, 'amount': amount}) 
					break

			line = fh.readline()
		fh.close()

'''
 Print the dictionaries.
 Structure the results into the format.
'''
print "["
for ia, artist in enumerate(dic.keys()):
	print '    {'
	print '        artist: '+artist+','
	print '        totalValue: \'USD '+"{:,}".format(dicValue[artist])+'\','
	print '        works: ['
	for i, work in enumerate(dic[artist]):
		print '            { title: '+work['title']+', currency: '+work['currency']+', amount: '+work['amount']+' }',
		if i != len(dic[artist]) - 1:
			print ','
		else:
			print
	print '        ]'
	if ia != len(dic.keys()) - 1:
		print '    },'
	else:
		print '    }'
print "]"
 # exit;

File: parser.js

var fs = require("fs")
var readline = require('readline')
var dirname ='data/2017-12-20/'

var files = fs.readdirSync('data/2017-12-20/')

fs.readdir(dirname, function(err, files) {
	if (err) {
		       return console.error(err);
		   }

	/* 
	* Define 2 object to save the data,
	* obj_list is the output JSON array,
	* obj_index saves the index of the artist in the obj_list.
	**/
	var obj_list = [];
	var obj_index = {};
	var index = 0;

	/*
	* Parse the html documents,
	* open the documents in the loop.
	**/
	const htmlFiles = files.filter(el => /\.html$/.test(el))
  	htmlFiles.forEach(file => 
  	{
    	var lines = require('fs').readFileSync(dirname+file, 'utf-8').split('\n');

    	var num = 1;
    	var artist, title, value, currency, amount;

    	/* search line by line */
    	lines.some(line => 
    	{
    		/* search the artist name in a html file */
			if (num == 1) 
			{
				reobj = /<h3 class='artist'>(.*?)( \(.*\))*<\/h3>/.exec(line);
				if (reobj) 
				{
					artist = '\''+reobj[1]+'\''
					/* initialize the object of the artist */
					if (!obj_index.hasOwnProperty(artist)) 
					{
						/* the artist's works are saved as a list */
						obj_list.push( {'artist': artist, 'totalValue': 0, 'works':[]} );
						/* saved the artist's index */
						obj_index[artist] = index;
						index++;
					}
					num++;
				}
			}
				

			/* search the title of the work in a html file */
			else if (num == 2) 
			{
				reobj = /<h3>(.*?)<\/h3>/.exec(line);
				if (reobj) 
				{
					title = '\''+reobj[1]+'\'';
					num++; 
				}
			}

			/* 
			* search the price of the work in a html file,
			* each work is saved as a object
			**/
			else if (num == 3) 
			{
				reobj = /<div><span class='currency'>(.*?)<\/span><span>(.*?)<\/span><\/div>/.exec(line);
				if (reobj) 
				{
					currency = '\''+reobj[1]+'\'';
					amount = '\''+reobj[2]+'\'';
					/* convert string to integer */
					value = parseInt(reobj[2].replace(/,/g, ''));	
					/* convert GBP to USD */
					if (currency == '\'GBP\'')	
					{
						value *= 1.34;
						currency = '\'USD\'';
					}
					/* calculate the totalValue */
					obj_list[obj_index[artist]].totalValue += value; 
					obj_list[obj_index[artist]].works.push({'title': title, 'currency': currency, 'amount': amount}) 
				}
			}
			return num === 4;
		});
  	});

  	/*
  	* Print the JSON array in the console.
  	* Structure the results into the format.
  	**/
  	obj_list.forEach(el => 
  	{
  		var value = el.totalValue.toLocaleString();
  		var value_string = '\'USD ' + value + '\'';
  		el.totalValue = value_string;
  	})
  	str = JSON.stringify(obj_list, null, 4);
	console.log(str.replace(/\"/g, ""));
})